### ISSUE TITLE
Addition error

### ISSUE THREAD

**Issue by wilbur-t** [JuliaDiff/ForwardDiff.jl#115]
Hello,

I received the following error while using ForwardDiff:

```
ERROR: Oops - +(::ForwardDiff.GradientNumber{0,Int64,Tuple{}}, ::ForwardDiff.GradientNumber{1,Real,Tuple{Real}}) should never have been called.
It was defined to resolve ambiguity, and was supposed to
fall back to a more specific method defined elsewhere.
Please report this bug to ForwardDiff.jl's issue tracker.
 in error at error.jl:21
```

As requested, I'm reporting the error.

This code reproduces the error:

```
using ForwardDiff

# Construct function
x    = Array{Real}(1)
x[1] = 1
f(theta) = sum(exp(x.*theta),1)

# Test that the function works
f([0.0])

# Test automatic derivative
g = ForwardDiff.gradient(f)
g([0.0])
```

Weird, huh!

I presumed that the issue might relate to the abstract array. However if you edit the line `x    = Array{Real}(1)` to `x    = Array{Float64}(1)` in the above code, a concrete array produces a different error.

Thanks in advance for the fix -- this seems like an important piece of software.

**Comment by jrevels** [comment]
Hi, thanks for the report!

I haven't totally figured this out yet, but I think it stems from within `sum` where an accumulation variable might be initializing with the wrong type. Probably a promotion/conversion error on our part.

Some tips that might help in the mean time:

- Changing `sum(exp(x.*theta),1)` to `sum(exp(x.*theta))` seems to make the whole thing work (and should be equivalent to what you wrote if `exp(x.*theta)` is always a vector).
- Maybe you simplified this example for the bug report, but if it's really the case that your input is always 1 element, you might want to use scalars and `ForwardDiff.derivative` instead. 
- Once again, this might just be a simplification made for the example, but Julia won't be able to optimize `f` with the type of `x` unless `x` is declared constant (or better, made a parameter to the function).

**Comment by wilbur-t** [comment]
Thanks for the reply, and for the tips. Yeah the example is (very!) simplified, and my sum is usually over the second dimension of a matrix. Replacing `sum(x[1:T-1,:],2)` with the comprehension `[sum(x[t,:]) for t in 1:T-1]` allows the automatic differentiation to work, though function evaluations now take roughly twice as long (I guess because the comprehension results in an `Array{Any,1}` while the original sum results in an `Array{Float64,2}`.)

By the way -- even using the comprehension, the automatic differentiation takes roughly a fifth of the time of finite-differences -- brilliant!

**Comment by jrevels** [comment]
This was fixed by #102. Since `f` is from `Array` to an `Array`, you need to use `ForwardDiff.jacobian` instead of `ForwardDiff.gradient`:

```julia
julia> ForwardDiff.jacobian(f, [0.0])
1x1 Array{Float64,2}:
 1.0
```
